### DISCUSSION TITLE
Debugging Parcel2 bundler plugins

### DISCUSSION THREAD

**Discussion by imakeinterfaces** [parcel-bundler/parcel#5455] [General]
I'm trying to add my own bundler plugin. I've followed the directions from the documentation and created a new plugin which looks something like this:

```
import { Bundler } from "@parcel/plugin";

export default new Bundler({  
  async bundle({ graph }) {
    // do stuff
  }
}
```
I'm not able to debug the process that runs the bundler initially, which is not making this any easier. In the past we would have used something like:
```
 bundler.on('buildStart', () => {
  // do stuff
 });
```

I'm assuming this is deprecated and that the async bundle function is called at buildStart. Can anyone confirm that this is the case?

**Comment by mischnic** [comment]
The bundler is executed on the main thread.

Are you sure your bundler plugin is actually the one that's being used?

**Reply by imakeinterfaces** [reply]
@mischnic How can I determine if this is the case? Also, do i need to add something to my .parcelrc file or somewhere else?

**Reply by mischnic** [reply]
> How can I determine if this is the case? 

console.log?
No bundles or some error if you have an empty function as in your original post

> i need to add something to my .parcelrc file or somewhere else?

You need something like this for a .parcelrc
```json
{
    "extends": "@parcel/config-default",
    "bundler": "parcel-bundler-thingy"
}
```

**Reply by imakeinterfaces** [reply]
@mischnic Thanks very much, this has worked. I'm still not able to see console logs in the normal parcel build mode but I can see that my bundle code is running. Now I have 2 more questions.

1. When i build with my new plugin and run my server, I now get a page which says `404
Page not found` which worked prior to adding my plugin.

2. What should be placed in the optimize function?

**Reply by mischnic** [reply]
Have you actually added code to do bundling in your plugin? Otherwise there will simply be no bundle and the server has nothing to serve.

bundle and optimize simply run in that order. At the moment, there is no difference between the two. So you could just do everything in bundle.

**Reply by imakeinterfaces** [reply]
@mischnic Yes i have code which is adding env vars to process.env. This plugin will be used to provide vars from the envkey library (https://www.npmjs.com/package/envkey) to the application. My app builds fine without the plugin enabled. What do I need to do to get the plugin to work alongside the built in bundler?

**Reply by mischnic** [reply]
I have a feeling the bundler plugin doesn't do what you think it does: https://v2.parceljs.org/plugin-system/overview/#complete-list-of-plugin-types-(in-a-somewhat-correct-order)

**Reply by imakeinterfaces** [reply]
You may be correct, what I do know is that the parcel v1 version of this plugin used the bundler.on event to handle the work of injecting the env vars. You can take a look at that here: https://github.com/dailybeast/parcel-envkey-plugin/blob/master/src/index.js Is there another way to use this event?

**Reply by mischnic** [reply]
The only way to set what `process.env` is programmatically is via the JS api: https://v2.parceljs.org/features/parcel-api/ (there an [`env` property](https://v2.parceljs.org/plugin-system/api/#InitialParcelOptions))

**Reply by imakeinterfaces** [reply]
@mischnic Thank you this seems to provide what I need. Unfortunately when I use the minimal example you provided, and add  
```    
serve: {
      port: 8080,
      distDir: '../dist'
}
```
to this config i'm now getting an error `@parcel/reporter-dev-server: Could not send success event to devserver because it does not exist.` This seems to be a problem reported elsewhere unfortunately.

**Reply by imakeinterfaces** [reply]
I will add this as another discussion as it no longer relates to the title of this one.
